File: typescript/src/model/specific_properties.ts
import { SpecificProperty } from './specific_property';

///class for the specificProperties property of RalObjects and RalMethods
export class SpecificProperties {
    private _specificProperties: Record<string, SpecificProperty>;

    constructor(specificProperties: Record<string, SpecificProperty>) {
        this._specificProperties = specificProperties;
    }

    getValueOf(key: string): any | null {
        const specificProperty = this._specificProperties[key];
        return specificProperty?.value ?? null;
    }

    containsKey(key: string): boolean {
        return key in this._specificProperties;
    }

    getProperty(key: string): SpecificProperty | null {
        const specificProperty = this._specificProperties[key];
        return specificProperty ? new SpecificProperty(specificProperty.key, specificProperty.value, specificProperty.unit ?? null) : null;
    }

    get map(): Record<string, SpecificProperty> {
        return { ...this._specificProperties };
    }

    /// returns the number of specific properties
    get count() {
        return Object.keys(this._specificProperties).length;
    }

    ///returns the specific properties as an array of maps
    toMaps(): Record<string, SpecificProperty>[] {
        return Object.values(this._specificProperties).map((value) => value.toMap());
    }

    static fromMaps(data: any): SpecificProperties {
        if (data === null || data === undefined) {
            return new SpecificProperties({});
        }

        if (!Array.isArray(data)) {
            throw new Error(`data must be an array, but was ${typeof data}`);
        }

        const specificProperties: Record<string, SpecificProperty> = {};
        for (const item of data) {

            if (typeof item !== 'object' || item === null) {
                throw new Error(`item must be an object, but was ${typeof item}`);
            }
            const specificProperty = SpecificProperty.fromMap(item);
            specificProperties[specificProperty.key] = specificProperty;
        }

        return new SpecificProperties(specificProperties);
    }
}

File: dart/lib/src/model/specific_properties.dart
import 'package:either_dart/either.dart';
import 'package:openral_core/src/cross/backend/parsing/parsing_error.dart';
import 'package:openral_core/src/model/specific_property.dart';
import 'package:openral_core/src/model_parser/specific_properties_parser.dart';

class SpecificProperties {
  final Map<String, SpecificProperty> _specificProperties;

  SpecificProperties(this._specificProperties);

  ///returns the value of the specific property with the given key or null if the property does not exist
  dynamic operator [](String key) => _specificProperties[key]?.value;

  bool containsKey(String key) {
    return _specificProperties.containsKey(key);
  }

  ///returns the specific property with the given key or null, if it does not exist
  SpecificProperty? get(String key) {
    return _specificProperties[key];
  }

  ///adds the given [SpecificProperty] to this [SpecificProperties], will override existing properties with the same key
  void set(SpecificProperty specificProperty) {
    _specificProperties[specificProperty.key] = specificProperty;
  }

  Map<String, SpecificProperty> get map => Map.fromEntries(_specificProperties.entries);

  ///returns [SpecificProperty]s as list of maps
  List<Map<String, dynamic>> toMaps() {
    return _specificProperties.values
        .map<Map<String, dynamic>>(
          (value) => value.toMap(),
        )
        .toList();
  }

  static Either<SpecificProperties, ParsingError> fromMaps(dynamic input) {
    final specificPropertiesParser = SpecificPropertiesParser();

    return specificPropertiesParser.parse(input);
  }
}
